edgemesh/pkg/proxy/proxy.go

```diff
@@ -0,0 +1,564 @@
+		ip := defaultNetworkPrefix + getSubNet(idx)
+		if err := vdev.AddIP(ip); err != nil {
+			vdev.DestroyDevice()
+			log.LOGGER.Errorf("[L4 Proxy] Add ip is failed : %s please checkout the env", err)
+			return
+		}
+		unused = append(unused, ip)
+	}
+	ipPoolSize = defaultIpPoolSize
+}
+
+// pollCallback process the connection from client
+func pollCallback(fd int32) {
+	ln := serve[fd].ln
+	serviceName := serve[fd].serviceName
+
+	conn, err := ln.Accept()
+	if err != nil {
+		return
+	}
+	getAndSetSocket(conn, false)
+	go startTcpServer(conn, serviceName)
+}
+
+// startTcpServer implement L4 proxy to the real server
+func startTcpServer(conn net.Conn, svcName string) {
+	portString := strings.Split(conn.LocalAddr().String(), ":")
+	// portString is a standard form, such as "172.17.0.1:8080"
+	localPort, _ := strconv.ParseInt(portString[1], 10, 32)
+	addr, err := doLoadBalance(svcName, localPort)
+	if err != nil {
+		log.LOGGER.Warnf("[L4 Proxy] %s call svc : %s encountered an error: %s", conn.RemoteAddr().String(), svcName, err)
+		conn.Close()
+		return
+	}
+	var proxyClient net.Conn
+	for retry := 0; retry < defaultTcpReconnectTimes; retry++ {
+		proxyClient, err = net.DialTimeout("tcp", addr.String(), defaultTcpClientTimeout)
+		if err == nil {
+			break
+		}
+	}
+	// Error when connecting to server ,maybe timeout or any other error
+	if err != nil {
+		log.LOGGER.Warnf("[L4 Proxy] %s call svc : %s to %s encountered an error: %s", conn.RemoteAddr().String(), svcName, addr.String(), err)
+		conn.Close()
+		return
+	}
+
+	ctk := &conntrack{
+		lconn: conn,
+		rconn: proxyClient,
+	}
+	log.LOGGER.Infof("[L4 Proxy] start a proxy server : %s,%s", svcName, addr.String())
+	go func() {
+		ctk.processServerProxy()
+	}()
+	go func() {
+		ctk.processClientProxy()
+	}()
+}
+
+// doLoadBalance implement the loadbalance function
+func doLoadBalance(svcName string, lport int64) (net.Addr, error) {
+	svc := strings.Split(svcName, ".")
+	namespace, name := svc[0], svc[1]
+	pods, err := metaClient.Services(namespace).GetPods(name)
+	if err != nil {
+		log.LOGGER.Errorf("[L4 Proxy] get svc error : %s", err)
+	}
+	// checkout the status of pods
+	runPods := make([]v1.Pod, 0, len(pods))
+	for i := 0; i < len(pods); i++ {
+		if pods[i].Status.Phase == v1.PodRunning {
+			runPods = append(runPods, pods[i])
+		}
+	}
+	// support random LB for the early version
+	rand.Seed(time.Now().UnixNano())
+	idx := rand.Uint32() % uint32(len(runPods))
+
+	hostIP := runPods[idx].Status.HostIP
+	st := addrByService.getAddrTable(svcName)
+
+	index := 0
+	for i, p := range st.ports {
+		if p == int32(lport) {
+			index = i
+			break
+		}
+	}
+	// kubeedge edgemesh support bridge net ,so, use hostport to access
+	tp := st.targetPort[index]
+	targetPort := int32(0)
+	for _, value := range runPods[idx].Spec.Containers {
+		for _, v := range value.Ports {
+			if v.ContainerPort == tp {
+				targetPort = v.HostPort
```

Comment:
we need to break the loop in case of match.

Reply:
ok，thanks